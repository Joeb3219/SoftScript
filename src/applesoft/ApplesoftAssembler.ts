import { OpcodeToApplesoftInstructionMap } from "./Applesoft.types";
import fs from "fs";

export class ApplesoftAssembler {
  private static readonly _MAX_LINE_NUMBER: number = 65_536;
  private static readonly _STARTING_ADDRESS: number = 2048;

  constructor(private readonly lines: string[]) {}

  private assembleLine(currentAddress: number, line: string): number[] {
    const parts = line.split(" ");
    const [lineNumberStr, operation, ...dataStr] = parts;
    const lineNumber = lineNumberStr ? parseInt(lineNumberStr) : -1;

    if (lineNumber < 0 || lineNumber >= ApplesoftAssembler._MAX_LINE_NUMBER) {
      throw new Error(`Line number is invalid: ${lineNumber}`);
    }

    const operationOpcode = Object.entries(
      OpcodeToApplesoftInstructionMap
    ).find((e) => e[1] === operation)?.[0];

    if (!operationOpcode) {
      throw new Error(`Operation is invalid: ${operation}`);
    }

    const data = dataStr
      .join(" ")
      .split("")
      .map((c) => c.charCodeAt(0));

    // 2 bytes for the next instruction address,
    // 2 bytes for the line number
    // 1 byte for the operation,
    // data.length bytes for the data,
    // and 1 byte for the 0x00 byte to end instruction.
    const numBytes = 2 + 2 + 1 + data.length + 1;
    const buffer = Buffer.alloc(numBytes);

    const nextInstructionAddress = currentAddress + numBytes + 1;

    // Write all of the data byte by byte, just like it was meant to be.
    buffer.writeUint16LE(nextInstructionAddress, 0);
    buffer.writeUint16LE(lineNumber, 2);
    buffer.writeUint8(parseInt(operationOpcode), 4);
    for (let i = 0; i < data.length; i++) {
      buffer.writeUint8(data[i], 5 + i);
    }
    buffer.writeUint8(0x00, numBytes - 1);

    return [...buffer];
  }

  assemble(): number[] {
    const lineBytes = this.lines.reduce<number[]>((state, line) => {
      const currentAddress =
        state.length + ApplesoftAssembler._STARTING_ADDRESS;
      const newBytes = this.assembleLine(currentAddress, line);

      return [...state, ...newBytes];
    }, []);

    // The last two bytes of the program are always 0x00.
    return [...lineBytes, 0x00, 0x00];
  }
}
