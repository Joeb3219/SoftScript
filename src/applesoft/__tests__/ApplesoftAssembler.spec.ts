import { ApplesoftAssembler } from "../ApplesoftAssembler";

describe("ApplesoftAssembler", () => {
    // These tests all write to a snapshot file that is hand-verified.
    describe("assemble", () => {
        it("should correctly handle REM instructions", () => {
            const assembler = new ApplesoftAssembler([
                `1 REM Eat "your" = vegetables`,
            ]);
            const result = assembler.assemble();
            expect(result).toMatchSnapshot();
        });

        it("should correctly handle string assignment", () => {
            const assembler = new ApplesoftAssembler([
                `1 LET X$ = "some value"`,
            ]);
            const result = assembler.assemble();
            expect(result).toMatchSnapshot();
        });

        it("should correctly handle multiple statements", () => {
            const assembler = new ApplesoftAssembler([
                `1 LET X$ = "some value"`,
                `2 PRINT X$`,
                `3 LET Y$ = X$ + "some other test"`,
                `4 PRINT Y$`,
                `5 GOTO 1`,
            ]);
            const result = assembler.assemble();
            expect(result).toMatchSnapshot();
        });
    });
});
