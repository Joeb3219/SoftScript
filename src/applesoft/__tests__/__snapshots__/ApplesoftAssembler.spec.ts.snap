// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplesoftAssembler assemble should correctly handle REM instructions 1`] = `
[
  31,
  8,
  1,
  0,
  178,
  32,
  69,
  97,
  116,
  32,
  34,
  121,
  111,
  117,
  114,
  34,
  32,
  61,
  32,
  118,
  101,
  103,
  101,
  116,
  97,
  98,
  108,
  101,
  115,
  0,
  0,
  0,
]
`;

exports[`ApplesoftAssembler assemble should correctly handle multiple statements 1`] = `
[
  22,
  8,
  1,
  0,
  170,
  88,
  36,
  208,
  34,
  115,
  111,
  109,
  101,
  32,
  118,
  97,
  108,
  117,
  101,
  34,
  0,
  30,
  8,
  2,
  0,
  186,
  88,
  36,
  0,
  59,
  8,
  3,
  0,
  170,
  89,
  36,
  208,
  88,
  36,
  200,
  34,
  115,
  111,
  109,
  101,
  32,
  111,
  116,
  104,
  101,
  114,
  32,
  116,
  101,
  115,
  116,
  34,
  0,
  67,
  8,
  4,
  0,
  186,
  89,
  36,
  0,
  74,
  8,
  5,
  0,
  171,
  49,
  0,
  0,
  0,
]
`;

exports[`ApplesoftAssembler assemble should correctly handle string assignment 1`] = `
[
  22,
  8,
  1,
  0,
  170,
  88,
  36,
  208,
  34,
  115,
  111,
  109,
  101,
  32,
  118,
  97,
  108,
  117,
  101,
  34,
  0,
  0,
  0,
]
`;
